pysrc/bytewax/run.py

```diff
@@ -0,0 +1,89 @@
+import argparse
+import pathlib
+
+from bytewax.recovery import SqliteRecoveryConfig
+
+from .bytewax import spawn_cluster, cluster_main
+
+__all__ = [
+    "cluster_main"
+]
+
+
+def _parse_args():
+    parser = argparse.ArgumentParser(
+        prog="python -m bytewax.run", description="Run a bytewax dataflow"
+    )
+    parser.add_argument("file_path", metavar="FILE_PATH", type=pathlib.Path)
+    run = parser.add_argument_group("Run options")
+    run.add_argument(
+        "-d",
+        "--dataflow-builder",
+        type=str,
+        default="get_flow",
+        help="Name of the function that returns the Dataflow.",
+    )
+    run.add_argument(
+        "--dataflow-args",
+        type=str,
+        nargs="*",
+        help="Args for the dataflow builder function.",
+    )
+    scaling = parser.add_argument_group(
+        "Scaling",
+        "You should use either '-p/-w' to spawn multiple processes on this same machine,"
+        " or '-i/-a' to spawn a single process on different machines",
+    )
+    scaling.add_argument(
+        "-p",
+        "--processes",
+        type=int,
+        help="Number of separate processes to run",
+    )
+    scaling.add_argument(
+        "-w",
+        "--workers-per-process",
+        type=int,
+        help="Number of workers for each process",
+    )
+    scaling.add_argument("-i", "--process-id", type=int, help="Process id")
+    scaling.add_argument(
+        "-a", "--addresses",
+        action="append",
+        help="Addresses of other processes"
+    )
+
+    # Config options for recovery
+    recovery = parser.add_argument_group("Recovery")
+    recovery.add_argument(
+        "--sqlite-directory",
+        type=pathlib.Path,
+        help="Passing this argument enables sqlite recovery in the specified folder",
+    )
+    recovery.add_argument(
+        "--epoch-interval",
+        type=int,
+        default=10,
+        help="Number of seconds between state snapshots",
+    )
+
+    args = parser.parse_args()
+    if (args.processes is not None or args.workers_per_process is not None) and (
+        args.process_id is not None or args.addresses is not None
+    ):
+        parser.error("Can't use both '-w/-p' and '-a/-i'")
+    return args
+
+
+def _main():
+    kwargs = vars(_parse_args())
+    sqlite_directory = kwargs.pop("sqlite_directory")
+    recovery_config = None
+    if sqlite_directory:
+        recovery_config = SqliteRecoveryConfig(sqlite_directory or "./")
+    kwargs["recovery_config"] = recovery_config
+    spawn_cluster(**kwargs)
```

Comment:
I'm a little confused here. What is the hierarchy of execution interfaces? Are some of these parts vestigial? It feels like another layer of abstraction was added, and did not simplify the hierarchy.

It seems like you have the CLI here delegating to `spawn_cluster` which then delegates to `DataflowRunner` but that has multiple modes, one for each kind of execution env? What is the purpose of `run_main` and `cluster_main` then? What does the extra layer of abstraction provided by `DataflowRunner` provide?

I would have assumed that (maybe other than marking `run_main` and `cluster_main` as "private" or "testing" in some way), we could keep `run_main` and `cluster_main` basically untouched, delete `spawn_cluster`, and then all of the logic of parsing command line args to determine execution mode, doing the dynamic `my_dataflow.py` importing, and possibly spawning subprocesses would all happen in just this script. I'm not totally sure I see the utility of the `DataflowRunner` and changing `run_main` and `cluster_main`.

Reply:
Yes, `DataflowRunner` itself added an extra layer of abstraction that in the end wasn't really adding anything, I think I just got excited with the refactoring :D I removed it, but I left the `WorkerRunner` struct, since that helps us with the bunch of arguments we were passing around to different functions. `run_main` and `cluster_main` are now back to what they were, we lose the ability to raise an exception if they are ran during import though. `spawn_cluster` is the new multiprocess runner. It's implemented on the Rust side just to have a better handling of possible errors and better control of some other aspects like running the webserver, but it can be moved here if we want all of this visible on the python side, the translation is almost 1:1 using `subprocess.Popen`. In fact in the first iteration it was all implemented here and I moved it to the Rust side on a later commit.